data/abi/abi_value.go

```diff
@@ -0,0 +1,304 @@
+		return Value{}, fmt.Errorf("static array make error: 0 argument passed in")
+	}
+	for i := 0; i < len(values); i++ {
+		if !values[i].ABIType.Equal(values[0].ABIType) {
+			return Value{}, fmt.Errorf("type mismatch: %s and %s",
+				values[i].ABIType.String(), values[0].ABIType.String())
+		}
+	}
+	return Value{
+		ABIType: MakeStaticArrayType(values[0].ABIType, uint16(len(values))),
+		value:   values,
+	}, nil
+}
+
+// MakeTuple takes an array of ABI values and returns an ABI tuple value.
+func MakeTuple(values []Value) (Value, error) {
+	if len(values) >= (1 << 16) {
+		return Value{}, fmt.Errorf("tuple make error: pass in argument number larger than 2^16")
+	}
+	tupleType := make([]Type, len(values))
+	for i := 0; i < len(values); i++ {
+		tupleType[i] = values[i].ABIType
+	}
+
+	castedTupleType, err := MakeTupleType(tupleType)
+	if err != nil {
+		return Value{}, err
+	}
+
+	return Value{
+		ABIType: castedTupleType,
+		value:   values,
+	}, nil
+}
+
+// MakeBool takes a boolean value and returns an ABI bool value.
+func MakeBool(value bool) Value {
+	return Value{
+		ABIType: MakeBoolType(),
+		value:   value,
+	}
+}
+
+// GetUint8 tries to retreve an uint8 from an ABI Value.
+func (v Value) GetUint8() (uint8, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 8 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return uint8(bigIntForm.Uint64()), nil
+}
+
+// GetUint16 tries to retrieve an uint16 from an ABI Value.
+func (v Value) GetUint16() (uint16, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 16 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return uint16(bigIntForm.Uint64()), nil
+}
+
+// GetUint32 tries to retrieve an uint32 from an ABI Value.
+func (v Value) GetUint32() (uint32, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 32 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return uint32(bigIntForm.Uint64()), nil
+}
+
+// GetUint64 tries to retrieve an uint64 from an ABI Value.
+func (v Value) GetUint64() (uint64, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 64 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return bigIntForm.Uint64(), nil
+}
+
+// GetUint tries to retrieve an big uint from an ABI Value.
+func (v Value) GetUint() (*big.Int, error) {
+	if v.ABIType.abiTypeID != Uint {
+		return nil, fmt.Errorf("value type unmatch")
+	}
+	bigIntForm := v.value.(*big.Int)
+	sizeThreshold := big.NewInt(0).Lsh(big.NewInt(1), uint(v.ABIType.bitSize))
+	if sizeThreshold.Cmp(bigIntForm) <= 0 {
+		return nil, fmt.Errorf("value is larger than uint bitSize")
+	}
```

Comment:
Is this checking that the value's size is bigger than its own belief about its bitsize?

Reply:
Yes, this checks is exactly checking sizeThreshold <= bigIntForm.

I want to make sure the value size is within the bitsize defined in type.